### PR TITLE
MAINT: Fix Box-Cox transform

### LINKED ISSUE #6873: 'stats.boxcox' return all same values
I applyed `stats.boxbox` to my data and the returned values are all the same, which seems really unreasonable!  it returned this same result in `scipy=0.18.1` and `scipy=0.17.1`. the  optimal `lambda` in my case is  -5.501196436791543.
your can download my data([data.txt](https://github.com/scipy/scipy/files/662791/data.txt)).
I also tried the `boxcox` function in `R` and it returned reasonable result. So  i think  the `boxcox_normmax`  function should do more test and should be used carefully.

### PR DESCRIPTION
**Pull request by omtinez**
* Fixes #6873
* Short-circuits MLE and Pearson in certain edge cases
* Adds test for Box-Cox transform for valid corner cases
* Adds test for Box-Cox transform for invalid corner cases

### CODE REVIEW

File: scipy/stats/tests/test_morestats.py
Comment (chrisb83) [inline scipy/stats/tests/test_morestats.py]: you need to check actual outcome against expected outcome using one of the `assert_...` functions

Comment (owahltinez) [reply scipy/stats/tests/test_morestats.py]: The actual outcome is irrelevant and to be considered undefined. These tests simply check that no warnings / errors are thrown. Without my changes, these tests fail.

Comment (omtinez) [reply scipy/stats/tests/test_morestats.py]: The actual outcome is irrelevant and to be considered undefined. These tests simply check that no warnings / errors are thrown. Without my changes, these tests fail.

Comment (ilayn) [reply scipy/stats/tests/test_morestats.py]: The output has to be checked regardless. Running the code successfully is already checking for exceptions.

Comment (omtinez) [reply scipy/stats/tests/test_morestats.py]: What should the output be? Right now, it is dependent on implementation details of the underlying functions all the way to the optimizer. I'm afraid that checking the output will be testing the implementation and not the specification.

The only thing I had in mind for these tests was to check for exceptions.

**Comment by omtinez** [comment]
> Just for the future reference, please make separate branches to work on new features and to send PRs such that your standard working repo does not interfere with the PRs you have submitted. Once a PR is merged you can safely delete that branch and keep working on other branches.

Makes sense. Fortunately this PR is the only one I have in flight right now. Thanks!

File: scipy/stats/tests/test_morestats.py
Comment (omtinez) [reply scipy/stats/tests/test_morestats.py]: > You can also call these private functions in the tests. 

Unfortunately, some of the functions are defined dynamically (def within def). Also, how would the higher level function be tested if I replace this test with several tests checking the private ones? Right now, there's a division by zero happening with the inputs chosen in the tests I added. Personally I'd prefer to make sure the test case passes with the higher level function rather than testing the private functions which are subject to change.

The tricky bit is that, in my opinion, the only reasonable thing to check for these test cases is that there's no error. The output might change if other private (shared) functions are changed, but that does not make the results of the higher level function less valid, and I'm trying not to test the implementation details.

**Comment by omtinez** [comment]
Friendly ping on this PR. Hoping to have this merged soon so I can remove workarounds that cause SciPy to crash in my code.

**Comment by chrisb83** [comment]
I currently don't have time to review in detail, just a few comments:

- can you describe the cases you are trying to fix in more detail and the rationale for the solution?
  - if the transformation returns non-finite values, returning nan looks fine
  - if the transformed values have MSE of zero, why set the MSE to 1e-21 in that case? if all values are constant, I guess there is no optimal lambda?

**Comment by omtinez** [comment]
Sure thing,

> can you describe the cases you are trying to fix in more detail and the rationale for the solution?

Intermediate values reach infinity due to division by zero or lack of enough precision in some scenarios. This simply means that the optimizer tried out a value that was not optimal, and letting the exception raise prevents the optimizer from searching further and returning a valid value.

> if the transformation returns non-finite values, returning nan looks fine

As far as I can tell, transformation should always return finite values and a valid lambda. In some edge cases the optimizer may not be able to find an optimal lambda (or there may not be an optimal lambda that makes sense, e.g. if all values are identical), but I don't think that returning nan is a valid output.

> if the transformed values have MSE of zero, why set the MSE to 1e-21 in that case? if all values are constant, I guess there is no optimal lambda?

Good question. I'm setting MSE to a very, very small number to prevent division by zero in the intermediate values, while letting the optimizer do its thing searching for a different value in whichever direction the result of that intermediate value tells it to go. If MSE is zero, it does not necessarily mean that there is no optimal lambda, only that the lambda that was tested in that iteration results in a transformed dataset that is equal to the input dataset. Unless I'm mistaken, that could be the most optimal way to transform the dataset, or there could be other lambdas yet to be tested by the optimizer that result in a better transformation.

Another thing that is being discussed, brought up by you initially and with concerns shared by @ilayn, is whether the output values should be checked. In the cases that I'm testing for, I'm only trying to make sure that the transformation succeeds. What transformation is expected in some corner cases -- some of which are identical to the corner cases that produce invalid intermediate values e.g. division by zero -- such as when all inputs are the same, is in my opinion undefined at the moment. Since the function in question is a statistical function, I would suggest to have a tighter definition of the expected output in those corner cases but I find that to be unrelated work to this PR which should be tested separately.

**Comment by chrisb83** [comment]
OK, thanks for the clarification, so llf should not raise an error, so the optimizer can continue to search for a solution.in one case (non-finite values), you return NaN, in the other case (MSE == 0), you tweak the MSE. Would NaN work in the latter case as well? (Might be more consistent)

File: scipy/stats/tests/test_morestats.py
Comment (chrisb83) [reply scipy/stats/tests/test_morestats.py]: All test cases you added are degenerate in the sense that there cannot be sensible value for lambda If I am not mistaken (single value input or all ones). In #6873, there were more "realistic" inputs and one could also test the outcome against R in that case. I think auch tests are better suited to check if the issue has been fixed.

**Comment by omtinez** [comment]
I think that returning NaN would not always work, specifically in the case where the input is already all equal values. By tweaking the MSE, we are making sure that there is at least one valid output. We could short-circuit if all inputs are the same, but I'm afraid that there may be contrived non-equal inputs where the solver is unable to find a single good solution if NaN is returned for MSE == 0.

Returning NaN for non-finite values does not appear to be a conscious choice, and I think in that case we would be better off hardening the internal functions and preventing all exceptions by checking inputs and values along the way; then keep searching for a valid lambda or return NaN when the search space has been exhausted.

But that would require a much larger change. For the time being, this fix appears to work for all the reports of Box-Cox misbehaving that I have seen on #6873.

File: scipy/stats/tests/test_morestats.py
Comment (owahltinez) [reply scipy/stats/tests/test_morestats.py]: As requested, I added `test_not_equal_output_values` using the input from one of the comments in the original isssue.

Comment (omtinez) [reply scipy/stats/tests/test_morestats.py]: As requested, I added `test_not_equal_output_values` with values from the original issue.

**Comment by omtinez** [comment]
Added an additional test as requested. Failing test seems to be flaky, unrelated to this change.

**Comment by omtinez** [comment]
Friendly ping. Is there anything left unaddressed that prevents this PR from being merged?

**Comment by tylerjereddy** [comment]
Just a mechanical comment -- you'll probably need to rebase to get the Appveyor failure resolved.

**Review by rgommers — CHANGES_REQUESTED** [review]
I have a few minor comments, but the main one is that the issue in https://github.com/scipy/scipy/issues/6873#issuecomment-411755758 isn't actually solved (even though this test passes):
```
In [17]: max(serie_bc) - min (serie_bc)
Out[17]: 2.9351283625800306e-09

In [18]: lmax
Out[18]: -2.2180716722493727
```

The outputs are not exactly equal, but `lmax` found is still outside the bracket, and the values are almost equal (relative difference on the order of 1e-9).

Using `stats.boxcox_normplot(x, -2.3, 2.1, plot=plt)` shows that `lmax=-2.21` is about right - the bracket in the issue is therefore misspecified. 

I'm actually not certain what the problem is here and what the expected response of the function should be - what should it do for inputs as in the `test_not_equal_output_values` function?

File: scipy/stats/tests/test_morestats.py
Comment (omtinez) [reply scipy/stats/tests/test_morestats.py]: Done

File: scipy/stats/morestats.py
Comment (omtinez) [reply scipy/stats/morestats.py]: Changed the arbitrary precision check to `np.unique`; to my surprise it seems to work just fine. I wouldn't be surprised if numpy is just doing something similar behind the scenes.

`np.nan` should not be returned here because it messes with the optimizer (and tests appear fail). I have not tested what the behavior would be when combined with gh-9562 but I would recommend against merging this change and that one simultaneously since they may affect each other. Although the tests I wrote for this change should be sufficient to tell whether a merge with gh-9562.

File: scipy/stats/morestats.py
Comment (omtinez) [reply scipy/stats/morestats.py]: Done

**Comment by omtinez** [comment]
Addressed requested changes. Test failures appear to be unrelated to this change as they can be seen in other PRs and I can confirm that all tests pass when I run them locally (using a fresh Docker image).

**Comment by rgommers** [comment]
@omtinez did you see my main comment in https://github.com/scipy/scipy/pull/9271#pullrequestreview-180631962 ("this issue isn't actually solved ...")?

**Review by rgommers — APPROVED** [review]

**Comment by rgommers** [comment]
> Maybe I'm not grasping that specific issue fully, but I think that there may be more than one problem being discusses in that thread. At least the issue that I experienced, which was being unable to find an lmax altogether, is fixed with this patch.

Probably, this PR does like an improvement. The `pearsonr` rewrite may help more though. I'm inclined to merge this but keep the issue open. Alternative is to merge the `pearsonr` rewrite first, and then revisit this. @WarrenWeckesser @chrisb83 thoughts?

File: scipy/stats/morestats.py
Comment (chrisb83) [inline scipy/stats/morestats.py]: how does that fit to the statement "the correlation is zero"?

**Comment by chrisb83** [comment]
The documentation of `boxcox_normmax` says the following:

> The starting interval for a downhill bracket search with optimize.brent. Note that this is in most cases not critical; the final result is allowed to be outside this bracket.

So the behaviour described by Ralf above appears to be fine.

Otherwise I cannot say much about the proposed solution, I don't fully understand the problem and I find the test cases confusing. For example, if a single value is supplied to boxcox, what should a meaningful output be? I would even say one should just raise an exception in that case.

Also, `assert_(max(output) - min(output) > 0)` does not appear to be a good conditon to check that the correct solution is found. Maybe use R as a benchmark (I could do that as a test in the next days.)?

I think the handling of pearsonr here needs to stay even if Warren's PR on pearson is fixed (i.e. raise an error if input is constant), whereas here the optimizer should just continue to search for a solution.

**Comment by omtinez** [comment]
> For example, if a single value is supplied to boxcox, what should a meaningful output be? I would even say one should just raise an exception in that case.

In that case where is the line drawn? What about lists of size two? I prefer not to raise an exception since that adds complexity to the users of that function (which also would need to be documented properly.)

> Also, assert_(max(output) - min(output) > 0) does not appear to be a good conditon to check that the correct solution is found. Maybe use R as a benchmark (I could do that as a test in the next days.)?

I'm open to suggestions. The current test was written to support that the quoted issue was fixed as requested.

**Comment by omtinez** [comment]
@rgommers will this PR be merged? Otherwise it should probably be closed as rejected since I won't be able to spend any more time on this.

**Comment by rgommers** [comment]
> @rgommers will this PR be merged? Otherwise it should probably be closed as rejected since I won't be able to spend any more time on this.

yes I think we'll merge it, possibly with some modifications. it's a useful contributions, please leave it open for now.

**Comment by WarrenWeckesser** [comment]
Don't merge this.  I don't think this is the right way to solve #6873.  I'm working on a different solution that reformulates the log-likelihood function to avoid the subtraction that leads to the loss of precision.  I'll submit an alternative pull request soon.

**Comment by WarrenWeckesser** [comment]
Sigh... when will I learn?  @omtinez, the *first* thing I should have said in my previous comment was *thank you for working on this!*  It is always great to have people who care enough about the project to submit bug fixes and other improvements.

In this case, though, I think I've worked out a better way to solve the problem.  I'll explain the details when I submit a new pull request.

**Comment by WarrenWeckesser** [comment]
My alternative solution is in https://github.com/scipy/scipy/pull/10072

**Comment by omtinez** [comment]
> Sigh... when will I learn? @omtinez, the _first_ thing I should have said in my previous comment was _thank you for working on this!_ It is always great to have people who care enough about the project to submit bug fixes and other improvements.

Thanks Warren, I appreciate your comment. It's this attitude that will make me much more likely to contribute to this project in the future (hopefully with higher-quality proposals).

@rgommers feel free to close as soon as @WarrenWeckesser's change is merged.

**Pull request closed without merging**
